cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,151 @@
+
+### RPC Methods
+
+Method : UpdateMappings
+
+Request : 
+```
+message FQDNMapping {
+    string FQDN = 1;
+    repeated bytes IPS = 2;
+    int32 TTL = 3;
+    bytes client_ip = 4;
+}
+```
+Response : 
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : UpdatesDNSRules
+
+_rpc UpdatesDNSRules(stream DNSPolicyRules) returns (Result){}_
+
+Request : 
+```
+message FQDNSelector {
+  string match_name = 1;
+  string match_pattern = 2;
+}
+
+message DNSPolicyRule {
+  uint32 port = 1;
+  string selector_string = 2;
+  repeated FQDNSelector port_rules= 3;
+  repeated string match_labels = 4;
+}
+
+message DNSPolicyRules {
+  uint64 endpoint_id = 1;
+  repeated DNSPolicyRule rules = 2;
+}
+```
+Response : 
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : GetEndpointMetadataByIP
+
+Request : 
+```
+message FQDNSelector {
+  string ip = 1;
+}
+```
+Response : 
+```
+message Result {
+    uint32 endpoint_id = 1;
+    uint32 identity = 2;
+}
+```
+
+### Tracking policy updates to SDP instances
+
+Since SDP can be deployed out of band and users can choose to completely disable built-in proxy to run multiple instances of SDP, agent should be prepared to handle multiple instances. In order to ensure all instances  have upto date policy revisions, agent will maintain a mapping of ACKed policy revision numbers against stream ID.
+
+
+## Impacts / Key Questions
+
+
+### Discovering endpoint metadata from SDP
+
+In order to enforce L7 DNS policy, SDP needs to translate IP address to endpoint ID and identity ID. The simplest option is to reuse the same gRPC stream and implement another method to get the mappings from cilium agent. Since the actual source of truth are the bpf maps. Should the SDP be allowed to directly read from bpf maps ?
+
+
+### Option 1: gRPC method
+
+#### Pros
+
+* Avoid interacting with low-level details. Simpler to maintain and stay upto date with any datapath changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+#### Cons
+
+* Reliance on gRPC call in the hot-path
+* In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
+
+### Option 2: Read from bpf maps
```

Comment:
This option definitely imposes some security permissions requirements on the SDP, because fetching BPF map information requires certain privileges (at least `CAP_BPF` on newer systems). Maybe if SDP also relies on `SO_REUSEPORT` then the privilege level is already fairly high? Not sure if this is specifically a comment about this option for the key question or whether we may want to have a security posture section to document the threat model and privilege expectations for this new component.

(This is not necessarily new; there are parallels with the Envoy implementation as well, which itself already looks up the IPCache for IP->Identity information for instance).

Reply:
This helps dnsproxy to reconcile without depending on Cilium Agent

Reply:
this helps solving dnsproxy to bootup without agent dependency